cips/cip-x.md

```diff
@@ -0,0 +1,77 @@
+| cip | XX (assigned by Editors) |
+| - | - |
+| title | Header Pruning for Light Nodes |
+| description | Mechanism to retain a fixed range of headers instead of whole history |
+| author | Hlib Kanunnikov ([@Wondertan](https://github.com/Wondertan)) |
+| discussions-to | <https://forum.celestia.org/t/cip-header-pruning-for-lns/1958> |
+| status | Draft |
+| type | Standards Track |
+| category | Data Availability |
+| created | 2025-03-21 |
+
+## Abstract
+
+Currently, every data availability (DA) node type synchronizes all historical headers starting from genesis (or other statically configured
+historical header) until the subjectively initialized head of the chain. We change that by adding a way to sync a
+constant size range of headers instead of the whole history.
+
+## Motivation
+
+Light nodes (LNs) store the whole history of ExtendedHeaders in their entirety, including ValidatorSet, Commit and data availability header (DAH). The motivation
+is to decrease LN storage requirements by retaining only a configurable window of recent headers.
+
+## Specification
+
+In order to reference the last stored header, we introduce a notion of **Tail** header of the chain, complementing
+**Head** header. Tail header height is estimated with [HeaderPruningWindow](#parameters), block time and Head height.
+
+- On the first start of a node, the Tail is retrieved from trusted peers alongside with the Head, performing subjective
+initialization.
+  - Once retrieved, Tail becomes the point of trust for the initializing node where everything beforehand assumed valid
+ (unless syncing from genesis)
+  - Head, on the other hand, is not trusted and becomes a synchronization target for the node.
+  - All the headers from Tail to Head are then retrieved and verified using signature-based forward verification.
+- As the chain progresses, nodes continuously re-estimate Tail and cut off stored headers beyond the new Tail, retaining
+roughly [HeaderPruningWindow](#parameters) amount of headers.
+
+The estimation of Tail height is done as follows:
+
+```go
+func estimateTail(head, tail Header, blockTime, pruningWindow time.Duration) (height uint64) {
+	estimatedRange = (now + pruningWindow - tail.Time()) / blockTime
+	estimatedCutoffHeight = min(tail.Height() + estimatedRange, head.Height())
+	return estimatedCutoffHeight
+}
+```
+
+## Parameters
+
+| Parameter           | Proposed value | Description                                   | Changeable via Governance |
+|---------------------|----------------|-----------------------------------------------|---------------------------|
+| HeaderPruningWindow | 30 days        | Defines the length of time to retain headers  | No                        |
+
+By default, HeaderPruningWindow (HPW) equals to SamplingWindow (SW) from [CIP-004](./cip-004.md). It can be modified per LN, but
```

Comment:
minor nit but unfortunately the literal `SamplingWindow` doesnt show up in CIP-004 even though it's defined implicitly

Reply:
What do you mean by "doesn't show up?"

Reply:
i mean ctrl-f for `SamplingWindow` shows nothing because its all "Sampling Window". a bit pedantic but just from a searchability perspective 😅 

also not seeing SamplingWindow in the https://celestiaorg.github.io/celestia-app/parameters.html

Reply:
> also not seeing SamplingWindow in the https://celestiaorg.github.io/celestia-app/parameters.html

similarly to comment above, this is because the parameters linked are specific to celestia-app. celestia-node parameters still live in code and not yet in the specs. cc @Wondertan and @walldiss here to get this implemented https://github.com/celestiaorg/celestia-node/tree/main/specs/src

Reply:
understood. let's link directly to CIP-4 for that from here to link together `HeaderPruningWindow` and `Sampling Window` and if we use `SamplingWindow` we should edit CIP-4

Reply:
> here to link together HeaderPruningWindow and Sampling Window

They are linked in Motivation and Parameters sections.

> if we use SamplingWindow we should edit CIP-4

Not sure why?

Reply:
> Not sure why?

Not necessary to do here now, I was proposing it as a way to resolve @ebuchman 's first comment